syntax = "proto3";

package cornserve.resource_manager;

import "common.proto";

service ResourceManager {
  // Deploy a new unit task
  rpc DeployUnitTask(DeployUnitTaskRequest) returns (DeployUnitTaskResponse);

  // Tear down a unit task
  rpc TeardownUnitTask(TeardownUnitTaskRequest) returns (TeardownUnitTaskResponse);

  // Health checking
  rpc Healthcheck(HealthcheckRequest) returns (HealthcheckResponse);

  // Scale up a unit task
  rpc ScaleUpUnitTask(ScaleUpUnitTaskRequest) returns (ScaleUpUnitTaskResponse);

  // Scale down a unit task
  rpc ScaleDownUnitTask(ScaleDownUnitTaskRequest) returns (ScaleDownUnitTaskResponse);
}

// Deploy a new task
message DeployUnitTaskRequest {
  // Task to deploy
  common.UnitTask task = 1;
}

message DeployUnitTaskResponse {
  // Success or failure
  common.Status status = 1;
}

// Tear down a task
message TeardownUnitTaskRequest {
  // Task to tear down
  common.UnitTask task = 1;
}

message TeardownUnitTaskResponse {
  // Success or failure
  common.Status status = 1;
}

// Health checking
message TaskManagerStatus {
  common.UnitTask task = 1;
  common.Status status = 2;
}

message HealthcheckRequest {}

message HealthcheckResponse {
  common.Status status = 1;
  repeated TaskManagerStatus task_manager_statuses = 2;
}

message ScaleUpUnitTaskRequest {
  // Task to scale up
  common.UnitTask task = 1;
  // Number of GPUs to add
  int32 num_gpus = 2;
}

message ScaleUpUnitTaskResponse {
  // Success or failure
  common.Status status = 1;
}

message ScaleDownUnitTaskRequest {
  // Task to scale down
  common.UnitTask task = 1;
  // Number of GPUs to remove
  int32 num_gpus = 2;
}

message ScaleDownUnitTaskResponse {
  // Success or failure
  common.Status status = 1;
}
